File: SwordToOffer/code/Test11.java
package ArithmeticCode.SwordToOffer.code;

/**
 * Created by Hy on 2019/11/25 9:45
 * <p>
 * 二进制数中1的个数
 */
public class Test11 {

    public static void main(String[] args) {
        int n = 1;
        for (int i = 0; i < Integer.MAX_VALUE; i++) {
            if (n == 0) break;
            System.out.println(" i = " + i);
            n = n << 1;
        }
    }

    public static class Solution {
        public int NumberOf1(int n) {
            int count = 0;
            int flag = 1;
            while (flag != 0) {
                if ((n & flag) != 0) {
                    count++;
                }
                flag = flag << 1;
            }
            return count;
        }

        public int NumberOf2(int n) {
            int count = 0;
            while (n != 0) {
                count++;
                //把最右边的1变成0
                n = (n - 1) & n;
            }
            return count;
        }
    }


}

File: kotlin/Test11.kt
package ArithmeticCode.kotlin

/**
 * Created by Hy on 2019/11/25 20:45
 */
fun main(){
    println(Test11().NumberOf1(8))
    println(Test11().NumberOf1(7))
    println(Test11().NumberOf1(3))

    println(Test11().NumberOf2(8))
    println(Test11().NumberOf2(7))
    println(Test11().NumberOf2(3))
}

class Test11 {

    fun NumberOf1(n:Int):Int {
        var count = 0
        var flag = 1
        while (flag != 0) {
            if (n.and(flag) != 0){
                count++
            }
            //移位操作
            flag = flag.shl(1)
        }
        return count
    }

    fun NumberOf2(n:Int):Int {
        var m = n;
        var count = 0;
        while (m != 0) {
            count++
            m = m.and(m-1)
        }
        return count;
    }

}
